### ISSUE TITLE
Keyword-spacing rule cannot parse typescript generated AST

### ISSUE THREAD

**Issue by soda0289** [eslint/eslint#8110]
**Tell us about your environment**
* **ESLint Version:**
master
* **Node Version:**
4.7.2
* **npm Version:**
4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?**
typescript-eslint-parser
**Please show your full configuration:**

```json


    "env": {
        "es6": true,
        "node": true,
        "mocha": true
    },
    "extends": "eslint:all",
    "parser": "typescript-eslint-parser",
    "parserOptions": {
        "sourceType": "module",
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "rules": {
    }


```

**What did you do? Please include the actual source code causing the issue.**

```ts

const foo =  symbol => 4;

@ffff() 
class Test {
    @ggg()
    static async values() {

    }
}

@ffff() 
class Test {
    @ggg()
    get values() {

    }

    @hhh()
    set values() {

    }
}


@eeeee
export function hhh(fff, @ff ee) {
}


@ttyy
export async function ffgghf(ff) {

}

@ffd
async function* gjgjg() {

}




```

**What did you expect to happen?**
It should parse

**What actually happened? Please include the actual, raw output from ESLint.**
An exception is thrown:

TypeScript will mark the first token of `symbol => 3;` as a keyword instead of as an identifier. This causes the function checkSpacingForFunction() to try and detect spacing around symbol which fails with: 
```
Cannot read property 'before' of undefined
TypeError: Cannot read property 'before' of undefined
    at checkSpacingBefore (/home/reyad/Workspace/eslint/lib/rules/keyword-spacing.js:276:40)
    at EventEmitter.checkSpacingForFunction (/home/reyad/Workspace/eslint/lib/rules/keyword-spacing.js:364:17)
    at emitOne (events.js:82:20)
    at EventEmitter.emit (events.js:169:7)
    at NodeEventGenerator.enterNode (/home/reyad/Workspace/eslint/lib/util/node-event-generator.js:39:22)
    at CodePathAnalyzer.enterNode (/home/reyad/Workspace/eslint/lib/code-path-analysis/code-path-analyzer.js:607:23)
    at CommentEventGenerator.enterNode (/home/reyad/Workspace/eslint/lib/util/comment-event-generator.js:98:23)
    at Controller.enter (/home/reyad/Workspace/eslint/lib/eslint.js:928:36)
    at Controller.__execute (/home/reyad/Workspace/eslint/node_modules/estraverse/estraverse.js:397:31)
    at Controller.traverse (/home/reyad/Workspace/eslint/node_modules/estraverse/estraverse.js:501:28)
```

Decorators are not supported by eslint but I think we can make the keyword-spacing rule work around them. The first token for a class method, with a decorator that starts with the keyword get, set, or async, is the at (@) symbol. This causes the keyword-spacing rule to fail with exception:

```
TypeError: Cannot read property 'before' of undefined
    at checkSpacingBefore (/home/reyad/Workspace/eslint/lib/rules/keyword-spacing.js:276:40)
    at checkSpacingAround (/home/reyad/Workspace/eslint/lib/rules/keyword-spacing.js:299:13)
    at EventEmitter.checkSpacingForProperty (/home/reyad/Workspace/eslint/lib/rules/keyword-spacing.js:503:17)
    at emitOne (events.js:82:20)
    at EventEmitter.emit (events.js:169:7)
    at NodeEventGenerator.enterNode (/home/reyad/Workspace/eslint/lib/util/node-event-generator.js:39:22)
    at CodePathAnalyzer.enterNode (/home/reyad/Workspace/eslint/lib/code-path-analysis/code-path-analyzer.js:607:23)
    at CommentEventGenerator.enterNode (/home/reyad/Workspace/eslint/lib/util/comment-event-generator.js:98:23)
    at Controller.enter (/home/reyad/Workspace/eslint/lib/eslint.js:928:36)
    at Controller.__execute (/home/reyad/Workspace/eslint/node_modules/estraverse/estraverse.js:397:31)
``` 

More information can be found here:
https://github.com/eslint/typescript-eslint-parser/issues/102

I'm not sure if this is the right way to solve these issues. The first problem, with symbol and arrow functions, is most likely a bug in typescript or something typescript-eslint-parser could handle. The second problem, with decorators, is something that can be handled better by the keyword-spacing rule. Typescript-eslint-parser could hide the decorator token, as babel-eslint does, but this isn't ideal since rules could be made for typescript decorators.
